File: holes/labs/wm-contract/runs/wm-09-predecessor-history-2026-09-16/k8-archive-tally.clj
(require '[clojure.java.io :as io] '[clojure.edn :as edn]
         '[futon2.aif.receipt-construction :as construction]
         '[futon2.aif.interpretation-evidence :as evidence]
         '[futon2.aif.evidence-manifest :as manifest])
(import '[java.nio.file Files])
(let [[root pinned output] *command-line-args*]
  (load-file pinned)
  (let [discover @#'construction/discover-closes!
        candidate @#'construction/closed-candidate!
        relevance @#'construction/relevance-evidence!
        admit @#'construction/validated-previous!
        coverage (discover [root])
        files (:files coverage)
        fresh "__wm09_diagnostic_fresh_target__"
        read-bytes #(Files/readAllBytes (.toPath (io/file %)))
        ;; Only rebind a manifest if every source is already in the verified
        ;; bounded copy and its literal digest matches. Otherwise stop safely.
        relocation
        (mapv (fn [path]
                (let [closed (edn/read-string (slurp path))
                      m (get-in closed [:payload :close-evidence-manifest])]
                  (if-not m {:close path :manifest :absent}
                    (let [entries (mapv (fn [e]
                                          (let [original (:source-path e)
                                                prefix "/home/joe/code/futon2/data/wm-full-loop/"
                                                _ (assert (.startsWith original prefix) (str "unavailable binding " original))
                                                dest (str root "/" (subs original (count prefix)))]
                                            (assert (.isFile (io/file dest)) (str "not in bounded copy " original))
                                            (assert (= (:sha256 e) (evidence/sha256 (read-bytes dest))))
                                            (assoc e :source-path dest))) (:entries m))
                          rebuilt (manifest/build-manifest {:entries entries :read-bytes read-bytes})]
                      (spit path (pr-str (assoc-in closed [:payload :close-evidence-manifest] rebuilt)))
                      {:close path :manifest :copy-path-rebound :verified-entries (count entries)})))) files)
        attempt (fn [f]
                  (try {:status :passed :value (f)}
                       (catch clojure.lang.ExceptionInfo e
                         {:status :refused :data (ex-data e)})))
        rows (mapv (fn [path]
                     (let [parsed (attempt #(candidate (io/file path)))]
                       (if (= :refused (:status parsed))
                         {:path path :phase :candidate-discovery :target-independent true :result parsed}
                         (let [c (:value parsed)]
                           (if (:non-construction? c)
                             {:path path :phase :non-construction :target-independent true
                              :status :excluded :provenance (:exclusion c)}
                             (do
                               (assert (not= fresh (:target c)))
                               {:path path :phase :target-relative :recorded-target (:target c)
                                :fresh-target {:requested-target fresh :result (attempt #(relevance c))}
                                :matching-target {:requested-target (:target c)
                                                  :result (attempt #(admit c [root]))}})))))) files)
        result {:source "858aa7f7" :scope :bounded-isolated-archive-diagnostic
                :coverage coverage :discovered (count files) :relocation relocation
                :rows rows :phase-counts (frequencies (map :phase rows))
                :fresh-result-counts (frequencies (keep #(get-in % [:fresh-target :result :status]) rows))
                :matching-result-counts (frequencies (keep #(get-in % [:matching-target :result :status]) rows))}]
    (assert (= 24 (count files)))
    (spit output (pr-str result))
    (prn (select-keys result [:discovered :phase-counts :fresh-result-counts :matching-result-counts]))))
(shutdown-agents)

File: holes/labs/wm-contract/runs/wm-09-predecessor-history-2026-09-16/k8-archive-tally.py
"""Bounded pinned archive copy; never invoke previous! on production roots."""
import hashlib, json, pathlib, shutil, subprocess, tempfile
repo = pathlib.Path('/home/joe/code/futon2')
r = repo / 'holes/labs/wm-contract/runs/wm-09-predecessor-history-2026-09-16'
census = pathlib.Path('/home/joe/code/p4ng/wm-walkthroughs/build-loop/closure/revision-21-inputs/archive-census.json')
data = json.loads(census.read_text())
assert data['count'] == len(data['files']) == 24
with tempfile.TemporaryDirectory(prefix='wm09-k8-archive-') as tmp:
    tmp = pathlib.Path(tmp)
    root = tmp / 'wm-full-loop'
    source_root = repo / 'data/wm-full-loop'
    records = []
    for row in data['files']:
        close = pathlib.Path(row['path'])
        rel = close.relative_to(source_root)
        assert rel.parts[:2] == ('archives', 'stop-line-2026-07-15')
        assert hashlib.sha256(close.read_bytes()).hexdigest() == row['sha256'], str(close)
        dest = root / rel.parent
        for source in close.parent.rglob('*'):
            assert not source.is_symlink(), str(source)
        shutil.copytree(close.parent, dest)
        for source in close.parent.rglob('*'):
            if source.is_file():
                copied = dest / source.relative_to(close.parent)
                assert hashlib.sha256(source.read_bytes()).digest() == hashlib.sha256(copied.read_bytes()).digest()
        records.append({'original': str(close), 'copy': str(root / rel), 'sha256': row['sha256']})
    pinned = tmp / 'receipt_construction.clj'
    pinned.write_bytes(subprocess.check_output(['git', 'show', '858aa7f7:src/futon2/aif/receipt_construction.clj'], cwd=repo))
    with (r / 'k8-archive-tally.log').open('w') as log:
        run = subprocess.run(['clojure', '-M:test', str(r / 'k8-archive-tally.clj'), str(root), str(pinned), str(r / 'K8-ARCHIVE-RESULTS.edn')], cwd=repo, stdout=log, stderr=subprocess.STDOUT)
    # Re-read pinned production closes to demonstrate no mutation by this run.
    for row in data['files']:
        assert hashlib.sha256(pathlib.Path(row['path']).read_bytes()).hexdigest() == row['sha256']
    (r / 'K8-ARCHIVE-COPY.json').write_text(json.dumps({'source_commit': '858aa7f7','census_sha256': hashlib.sha256(census.read_bytes()).hexdigest(),'verified_copied_closes': records,'exit': run.returncode,'production_close_hashes_unchanged': True,'temporary_copy_removed_on_exit': True}, indent=2)+'\n')
    assert run.returncode == 0
